model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -26,50 +31,314 @@ pub struct CommandData {
 
 /// Data received when a user fills in a command option.
 ///
-/// Note: user, channel, role, and mentionable option types will be returned as
-/// a [`String`] option here.
-///
 /// Refer to [the discord docs] for more information.
 ///
 /// [the discord docs]: https://discord.com/developers/docs/interactions/slash-commands#interaction-applicationcommandinteractiondataoption
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
-#[serde(untagged)]
-pub enum CommandDataOption {
-    String {
-        name: String,
-        value: String,
-    },
-    Integer {
-        name: String,
-        value: i64,
-    },
-    Boolean {
-        name: String,
-        value: bool,
-    },
-    SubCommand {
-        name: String,
-        #[serde(default)]
-        options: Vec<CommandDataOption>,
-    },
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CommandDataOption {
+    pub name: String,
+    pub value: CommandOptionValue,
 }
 
-impl CommandDataOption {
-    pub const fn kind(&self) -> &'static str {
-        match self {
-            CommandDataOption::String { .. } => "String",
-            CommandDataOption::Integer { .. } => "Integer",
-            CommandDataOption::Boolean { .. } => "Boolean",
-            CommandDataOption::SubCommand { .. } => "SubCommand",
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CommandOptionValue {
+    String(String),
+    Integer(i64),
+    Boolean(bool),
+    User(User, Option<InteractionMember>),
```

Comment:
Let's just hold the `UserId` here. Let the user decide whether they want to resolve it or not